perception/sub8_perception/include/sub8_perception/start_gate.hpp

```diff
@@ -1,67 +1,105 @@
 #pragma once
+
 #include <iostream>
 #include <string>
+#include <vector>
 
-#include "ros/ros.h"
+#include <boost/bind.hpp>
+#include <boost/date_time/posix_time/posix_time.hpp>
+#include <boost/thread.hpp>
 
+#include <cv_bridge/cv_bridge.h>
 #include <image_geometry/pinhole_camera_model.h>
 #include <image_transport/image_transport.h>
+#include <ros/ros.h>
+#include <std_srvs/SetBool.h>
 
-#include <cv_bridge/cv_bridge.h>
-#include <opencv2/highgui/highgui.hpp>
+#include <mil_vision_lib/cv_tools.hpp>
+
+#include <eigen_conversions/eigen_msg.h>
+#include <Eigen/Core>
+#include <Eigen/Dense>
+#include <Eigen/Geometry>
+#include <Eigen/StdVector>
 
 #include <boost/accumulators/accumulators.hpp>
-#include <boost/accumulators/statistics.hpp>
 #include <boost/circular_buffer.hpp>
 
-#include <std_srvs/SetBool.h>
-
-#include "sub8_msgs/BMatrix.h"
-#include "sub8_msgs/VisionRequest2D.h"
-
-using namespace boost::accumulators;
+#include <visualization_msgs/Marker.h>
 
 class Sub8StartGateDetector
 {
 public:
   Sub8StartGateDetector();
-  ~Sub8StartGateDetector();
-
-  // 76 inches x 39.5 inches
-  // 1.9304 x 1.0033
-  const double START_GATE_WIDTH = 1.9304;  // Meters
-  const double START_GATE_HEIGHT = 1.0033;
-  const double START_GATE_SIZE_TOLERANCE = 10000;
-
-  void imageCallback(const sensor_msgs::ImageConstPtr &msg, const sensor_msgs::CameraInfoConstPtr &info_msg);
-  bool requestStartGatePosition2d(sub8_msgs::VisionRequest2D::Request &req, sub8_msgs::VisionRequest2D::Response &resp);
-  bool requestStartGateEnable(std_srvs::SetBool::Request &req, std_srvs::SetBool::Response &resp);
-  bool requestStartGateDistance(sub8_msgs::BMatrix::Request &req, sub8_msgs::BMatrix::Response &resp);
-  void findGate(const sensor_msgs::ImageConstPtr &image_msg);
-
-  double getGateDistance();
-
-  ros::NodeHandle nh_;
-  bool running_;
-
+  mil_vision::ImageWithCameraInfo left_most_recent, right_most_recent;
```

Comment:
Did you try using ROSCameraStream? it makes a lot of things more convenient, adds buffer functionality. Heres a [simple demo](https://github.com/uf-mil/mil_common/blob/70af0ae7832b2863a4d6c7251fd20599fa6ace84/perception/mil_vision/src/camera_stream_demo.cpp).